File: cpp/fx-options-engine/src/implied_vol.cpp
#include "fxopt/implied_vol.hpp"

#include <algorithm>
#include <cmath>

#include "fxopt/garman_kohlhagen.hpp"
#include "brent.hpp"

namespace fxopt {

namespace {

// GK vega (dV/dsigma), same for calls and puts.
double gk_vega(double S, double K, double T, double r_d, double r_f,
               double sigma) {
    const double d1v = d1(S, K, T, r_d, r_f, sigma);
    return S * std::exp(-r_f * T) * norm_pdf(d1v) * std::sqrt(T);
}

}  // namespace

double implied_vol(double price, double S, double K, double T, double r_d,
                   double r_f, OptionType type, double tol, int max_iter) {
    const double p = phi(type);
    validate_inputs(S, K, T, r_d, r_f, 0.0);
    if (T <= 0.0) {
        throw std::invalid_argument("implied_vol requires T > 0");
    }
    detail::require_finite(price, "price");

    const double df_d = std::exp(-r_d * T);
    const double df_f = std::exp(-r_f * T);
    const double forward = S * df_f / df_d;
    const double lower = df_d * std::max(p * (forward - K), 0.0);  // sigma->0
    const double upper = p > 0.0 ? S * df_f : K * df_d;            // sigma->inf
    if (price < lower - 1e-14 || price > upper + 1e-14) {
        throw std::invalid_argument(
            "price " + std::to_string(price) +
            " outside no-arbitrage bounds [" + std::to_string(lower) + ", " +
            std::to_string(upper) + "]");
    }
    if (price - lower <= 1e-16 * std::max(1.0, lower)) {
        // Time value below double-precision resolution: vol unrecoverable,
        // return the sigma -> 0 limit (documented in docs/VALIDATION.md).
        return 0.0;
    }

    const auto objective = [&](double sig) {
        return gk_price(S, K, T, r_d, r_f, sig, type) - price;
    };

    // Newton with a moneyness-aware start (Brenner-Subrahmanyam flavoured).
    double sigma =
        std::max(0.05, std::sqrt(2.0 * std::abs(std::log(forward / K)) / T));
    const double lo = 1e-10;
    double hi = 10.0;
    for (int i = 0; i < max_iter; ++i) {
        const double diff = objective(sigma);
        if (std::abs(diff) < 1e-14) {
            return sigma;
        }
        const double vega = gk_vega(S, K, T, r_d, r_f, sigma);
        if (vega < 1e-12) {
            break;  // flat objective; Newton unreliable -> Brent
        }
        const double new_sigma = sigma - diff / vega;
        if (!(new_sigma > lo && new_sigma < hi)) {
            break;
        }
        if (std::abs(new_sigma - sigma) < tol) {
            return new_sigma;
        }
        sigma = new_sigma;
    }

    // Brent fallback on an expanding bracket.
    hi = 1.0;
    double f_hi = objective(hi);
    while (f_hi < 0.0 && hi < 50.0) {
        hi *= 2.0;
        f_hi = objective(hi);
    }
    if (f_hi < 0.0) {
        throw std::invalid_argument("implied vol > " + std::to_string(hi) +
                                    ": price " + std::to_string(price) +
                                    " unattainably high");
    }
    if (f_hi == 0.0) {
        // objective(hi) landed exactly on zero without ever going strictly
        // positive during expansion: deep ITM + long-dated + high vol drives
        // |d1|, |d2| large enough that N(d1)/N(d2) saturate to 0 or 1 in
        // double precision, so gk_price(sigma) is bit-identical to the
        // sigma -> inf bound for every sigma from the true root up to `hi`
        // (and beyond -- this is not a bracket, it is a flat plateau). Any
        // point in that plateau is an equally "valid" root of the floating
        // point objective, so accepting `hi` (an arbitrary artifact of the
        // doubling schedule) would silently return a vol that can be wrong
        // by whole vol points or more with no signal to the caller. This is
        // the upper-bound mirror of the near-`lower` short-circuit above;
        // unlike that case there is no finite limiting sigma to fall back
        // to (sigma -> infinity is not representable), so the honest
        // answer is that the vol is unrecoverable at this precision.
        throw std::invalid_argument(
            "price " + std::to_string(price) +
            " is within double-precision resolution of the sigma->inf "
            "bound " + std::to_string(upper) +
            "; implied volatility is unrecoverably large (vega has "
            "underflowed to zero in this regime -- see docs/VALIDATION.md)");
    }
    return detail::brentq(objective, lo, hi, tol, 200);
}

}  // namespace fxopt

File: rust/fx-options-engine/src/implied_vol.rs
//! Implied Garman–Kohlhagen volatility from a domestic-currency premium.
//!
//! Strategy: Newton–Raphson from a moneyness-aware initial guess (fast
//! quadratic convergence when vega is healthy), falling back to bracketed
//! Brent (guaranteed convergence) if Newton stalls or wanders outside the
//! no-arbitrage bracket.  Mirrors the Python
//! `fx_options.garman_kohlhagen.implied_vol` and the C++ engine exactly:
//! same bounds checks, same tolerances, same degenerate-price behaviour.

use crate::garman_kohlhagen::{d1 as d1_fn, gk_price};
use crate::{brentq, invalid, norm_pdf, require_finite, validate_inputs, FxResult, OptionType};

/// Absolute tolerance on the vol root (round trips hold to 1e-10).
const TOL: f64 = 1e-12;
/// Newton iteration budget before the Brent fallback.
const MAX_ITER: usize = 100;

/// GK vega (dV/dsigma), same for calls and puts.
fn gk_vega(s: f64, k: f64, t: f64, r_d: f64, r_f: f64, sigma: f64) -> FxResult<f64> {
    let d1v = d1_fn(s, k, t, r_d, r_f, sigma)?;
    Ok(s * (-r_f * t).exp() * norm_pdf(d1v) * t.sqrt())
}

/// Implied Garman–Kohlhagen volatility from a domestic-currency premium.
///
/// # Arguments
///
/// * `price` — observed premium, domestic ccy per unit foreign notional.
/// * `s`, `k`, `t`, `r_d`, `r_f` — as in [`crate::gk_price`]; requires
///   `t > 0`.
/// * `option_type` — call or put.
///
/// Returns the implied volatility (annualised) to absolute tolerance
/// 1e-12 (price round trips reproduce the input vol to better than
/// 1e-10 across the tested strike/vol grid).
///
/// A price whose time value is below double-precision resolution returns
/// `0.0` (the `sigma -> 0` limit; vol unrecoverable — documented in
/// docs/VALIDATION.md).
///
/// # Errors
///
/// [`crate::FxError::InvalidInput`] if the price violates the
/// no-arbitrage bounds `[discounted intrinsic on the forward, discounted
/// forward bound]` or `t = 0`.
///
/// ```
/// use fx_options_engine::{gk_price, implied_vol, OptionType};
/// let (s, k, t, rd, rf) = (1.10, 1.15, 0.5, 0.0425, 0.0290);
/// let px = gk_price(s, k, t, rd, rf, 0.0925, OptionType::Call).unwrap();
/// let iv = implied_vol(px, s, k, t, rd, rf, OptionType::Call).unwrap();
/// assert!((iv - 0.0925).abs() < 1e-10);
/// ```
pub fn implied_vol(
    price: f64,
    s: f64,
    k: f64,
    t: f64,
    r_d: f64,
    r_f: f64,
    option_type: OptionType,
) -> FxResult<f64> {
    let phi = option_type.phi();
    validate_inputs(s, k, t, r_d, r_f, 0.0)?;
    if t <= 0.0 {
        return invalid("implied_vol requires T > 0");
    }
    require_finite(price, "price")?;

    let df_d = (-r_d * t).exp();
    let df_f = (-r_f * t).exp();
    let forward = s * df_f / df_d;
    let lower = df_d * (phi * (forward - k)).max(0.0); // sigma -> 0 limit
    let upper = if phi > 0.0 { s * df_f } else { k * df_d }; // sigma -> inf
    if price < lower - 1e-14 || price > upper + 1e-14 {
        return invalid(format!(
            "price {price} outside no-arbitrage bounds [{lower}, {upper}]"
        ));
    }
    if price - lower <= 1e-16 * lower.max(1.0) {
        // Time value below double-precision resolution: vol unrecoverable,
        // return the sigma -> 0 limit (documented in docs/VALIDATION.md).
        return Ok(0.0);
    }

    let objective = |sig: f64| -> FxResult<f64> {
        Ok(gk_price(s, k, t, r_d, r_f, sig, option_type)? - price)
    };

    // Newton with a moneyness-aware start (Brenner-Subrahmanyam flavoured).
    let mut sigma = (2.0 * (forward / k).ln().abs() / t).sqrt().max(0.05);
    let lo = 1e-10;
    let hi_cap = 10.0;
    for _ in 0..MAX_ITER {
        let diff = objective(sigma)?;
        let vega = gk_vega(s, k, t, r_d, r_f, sigma)?;
        if vega < 1e-12 {
            // Flat objective: an absolute price tolerance here would be
            // meaningless (a wing option's vega can be ~1e-9, so even a
            // sub-ULP price residual implies a sigma error orders of
            // magnitude above `TOL` — this is what let short-dated wing
            // round trips return early ~1e-7 off).  Only accept if the
            // price residual is already at the double-precision noise
            // floor; otherwise hand off to bracketed Brent, which solves
            // in the sigma domain and is immune to the flat vega.
            if diff.abs() < 1e-14 {
                return Ok(sigma);
            }
            break;
        }
        let new_sigma = sigma - diff / vega;
        if !(new_sigma > lo && new_sigma < hi_cap) {
            break;
        }
        if (new_sigma - sigma).abs() < TOL {
            sigma = new_sigma;
            break;
        }
        sigma = new_sigma;
    }

    // Brent fallback.  The objective cannot fail inside [lo, hi_cap]:
    // sigma > 0, t > 0 and the inputs are validated.
    let obj_infallible = |sig: f64| {
        gk_price(s, k, t, r_d, r_f, sig, option_type)
            .map(|p| p - price)
            .unwrap_or(f64::NAN)
    };

    // First try a bracket centred tightly on the last Newton iterate and
    // expand it geometrically.  For flat-vega wings (short-dated, deep
    // ITM/OTM) the GK call/put formula loses precision to cancellation
    // (S e^{-r_f T} N(d1) - K e^{-r_d T} N(d2), both O(1), difference
    // O(price)); the objective is then a *staircase* in sigma with ULP-
    // sized flats, and Brent converges to the nearest edge of that flat
    // fastest, and with least residual bias, when seeded from a narrow
    // window around a value already close to the root — a wide bracket
    // approaches the same flat from farther away and can settle on the
    // wrong edge.  Falls back to the old wide search from 1.0 if the
    // narrow window never straddles a sign change (e.g. Newton broke out
    // on iteration 0 with a poor start).
    let mut width = (sigma * 1e-4).max(1e-6);
    while width < 2.0 {
        let blo = (sigma - width).max(lo);
        let bhi = (sigma + width).min(hi_cap);
        let flo = obj_infallible(blo);
        let fhi = obj_infallible(bhi);
        if flo == 0.0 {
            return Ok(blo);
        }
        if fhi == 0.0 {
            return Ok(bhi);
        }
        if flo.is_finite() && fhi.is_finite() && flo * fhi < 0.0 {
            return brentq(obj_infallible, blo, bhi, TOL, 200);
        }
        width *= 8.0;
    }

    let mut hi = 1.0;
    while obj_infallible(hi) < 0.0 && hi < 50.0 {
        hi *= 2.0;
    }
    if obj_infallible(hi) < 0.0 {
        return invalid(format!("implied vol > {hi}: price {price} unattainably high"));
    }
    brentq(obj_infallible, lo, hi, TOL, 200)
}
